1.segment_images/scripts/1.segment_cells_organoids.py

```diff
@@ -0,0 +1,262 @@
+import sys
+
+import matplotlib.pyplot as plt
+
+# Import dependencies
+import numpy as np
+import skimage
+import tifffile
+from cellpose import core
+from cellpose import io as cellpose_io
+from cellpose import models
+
+cellpose_io.logger_setup()
+import torch
+from cellpose.io import imread
+from PIL import Image
+from skimage import io
+
+use_GPU = torch.cuda.is_available()
+
+# set import path
+sys.path.append(str(pathlib.Path("../../utils/").resolve()))
+
+# check if in a jupyter notebook
+try:
+    cfg = get_ipython().config
+    in_notebook = True
+except NameError:
+    in_notebook = False
+
+print(in_notebook)
+
+
+# In[2]:
+
+
+if not in_notebook:
+    # set up arg parser
+    parser = argparse.ArgumentParser(description="Segment the nuclei of a tiff image")
+
+    parser.add_argument(
+        "--input_dir",
+        type=str,
+        help="Path to the input directory containing the tiff images",
+    )
+    parser.add_argument(
+        "--window_size", type=int, help="Size of the window to use for the segmentation"
+    )
+    parser.add_argument(
+        "--clip_limit",
+        type=float,
+        help="Clip limit for the adaptive histogram equalization",
+    )
+
+    args = parser.parse_args()
+    window_size = args.window_size
+    clip_limit = args.clip_limit
+    input_dir = pathlib.Path(args.input_dir).resolve(strict=True)
+
+else:
+    input_dir = pathlib.Path("../../data/z-stack_images/C4-2/").resolve(strict=True)
+    window_size = 3
+    clip_limit = 0.1
+
+mask_path = pathlib.Path(f"../processed_data/{input_dir.stem}").resolve()
+mask_path.mkdir(exist_ok=True, parents=True)
+
+
+# ## Set up images, paths and functions
+
+# In[3]:
+
+
+image_extensions = {".tif", ".tiff"}
+files = sorted(input_dir.glob("*"))
+files = [str(x) for x in files if x.suffix in image_extensions]
+
+
+# In[4]:
+
+
+# find the cytoplasmic channels in the image set
+for f in files:
+    if "405" in f:
+        nuclei = io.imread(f)
+    elif "488" in f:
+        cyto1 = io.imread(f)
+    elif "555" in f:
+        cyto2 = io.imread(f)
+    elif "640" in f:
+        cyto3 = io.imread(f)
+    else:
+        print(f"Unknown channel: {f}")
+
+# pick which channels to use for cellpose
+cyto = skimage.exposure.equalize_adapthist(cyto2, clip_limit=clip_limit)
+
+
+original_nuclei_image = nuclei.copy()
+original_cyto_image = cyto.copy()
+
+original_nuclei_z_count = nuclei.shape[0]
+original_cyto_z_count = cyto.shape[0]
+
+
+# In[5]:
+
+
+# make a 2.5 D max projection image stack with a sliding window of 3 slices
+image_stack_2_5D = np.empty((0, cyto.shape[1], cyto.shape[2]), dtype=cyto.dtype)
+for image_index in range(cyto.shape[0]):
+    image_stack_window = cyto[image_index : image_index + window_size]
+    if not image_stack_window.shape[0] == window_size:
+        break
+    # max project the image stack
+    image_stack_2_5D = np.append(
+        image_stack_2_5D, np.max(image_stack_window, axis=0)[np.newaxis, :, :], axis=0
+    )
+
```

Comment:
Looking at your other codes, this may be similar or the same for loops that I have been seeing. Would it be best to make this as a function? Making the code much easier and much more modular in order to prevent repeating yourself? By looking at this, a good template for this would be something like this:

```suggestion
def create_sliding_window_stack(image, window_size):
    stack_2_5D = np.empty((0, image.shape[1], image.shape[2]), dtype=image.dtype)
    for idx in range(image.shape[0]):
        window = image[idx : idx + window_size]
        if window.shape[0] != window_size:
            break
        stack_2_5D = np.append(
            stack_2_5D, np.max(window, axis=0)[np.newaxis, :, :], axis=0
        )
    return stack_2_5D
```

Assuming that all inputs are the same despite which channel you are looking at. (I changed the varaibles to make it simpler and general using this as a function)

Reply:
I agree with you again. Each of these analyses will be turned into a series of scripts (similar to pycytominer) but not just yet. I will be using everything in a notebook first for a more experimental design!